players/g3_player.py

```diff
@@ -207,51 +208,107 @@ def play(self, unit_id: List[List[str]], unit_pos: List[List[Point]], map_states
 
             scout_ids = np.arange(self.num_scouts)
 
-            #start = time.time()
+            start = time.time()
             defense_moves = self.push(scout_ids)
-            #print('Defense:', time.time()-start)
+            self.debug(f'Defense: {time.time()-start}s')
             
-            #start = time.time()
+            start = time.time()
             offense_moves = self.move_scouts(scout_ids)
-            #print('Offense:', time.time()-start)
+            self.debug(f'Offense: {time.time()-start}s')
 
             return offense_moves + defense_moves
 
-
     def order2coord(self, order: Tuple[float, float]) -> Tuple[float, float]:
         """Converts an order, tuple of (dist2homebase, angle), into a coordinate."""
         dist, angle = order
         x = self.homebase[0] + dist * math.cos(angle)
         y = self.homebase[1] + dist * math.sin(angle)
         return (x, y)
 
-    def explore(self, scout_unit, ally_units):
-        # change to random selection to speed up
-        enemy_clusters = self.enemy_units[np.random.choice(np.arange(self.enemy_units.shape[0]), min(self.enemy_units.shape[0], 50), replace=False)]
-        ally_clusters = ally_units[np.random.choice(np.arange(ally_units.shape[0]), min(ally_units.shape[0], 15), replace=False)]
-
+    def get_border(self):
+        """Get border of our territory"""
+        # trace along x axis to find the starting point
+        if self.us < 2: # 0, 1
+            for i in range(100):
+                if self.map_states[i, 99*self.us] != self.us:
+                    pt = (i-1, 99*self.us)
+                    break
+        else: # 2, 3
+            for i in range(100):
+                if self.map_states[99-i, 99*(3-self.us)] != self.us:
+                    pt = (99-i+1, 99*(3-self.us))
+                    break
```

Comment:
If I understand correctly, you are trying to find the map edge that are part of a player's border here.

It seems that here player index 1 (i.e player 2) is at the top right, but I think the simulator puts player on bottom left?

Reply:
(0, 99) is indeed bottom left...

Reply:
gotcha, I think I confused the position of the element in the matrix with where it is on the map.

Reply:
When i wrote the code, I still think about it with the matrix and just flip player index 1 and 3 in my head. Doesn't change anything, unless moves are involved